# CSCI 4580/5580 - Data Science – Fall 2020
## Assignment 3: kNN, Naive Bayes and Text Feature Selection

#### Student Full Name: Sharvita Paithankar
#### Student ID: 108172438

### Deliverables
Complete all the exercises below and turn in a write-up in the form of a Jupyuter notebook, that is, an **.ipynb file**.
The write-up should include your code, answers to exercise questions, and plots of results.
You will submit your assignment online as an attachment, through Canvas under Assignment 3. 

You have to use this notebook and fill in answers inline.
Don't forget to include answers to questions that ask for natural language responses, i.e., in English, not code!

Please do not add or remove any cells (you can add cells while doing the homework but remove them before submission).

### Need Help? 
In case you need help with this assignment, please ask your questions on the designated channel for Assignment 3 on Microsoft Teams. This way, you can receive assistance from the TA’s or your classmates (<b>DO NOT share your solution</b>) [Recommended]. 
<br/>If you need to include your solution in your question, please create a new chat with <b>all the TA’s</b> (in the new chat, add all the TA’s rather than sending your question to each TA separately).

### Collaboration
This assignment is to be done individually. Everyone should obtain hands-on experience in this course. You are free to discuss course material with fellow students, and we encourage you to use the resources in the Internet for your understanding, but the work you turn in, including all code and answers, must be your own work.

Let's Get Started!
===
## Overview

Supervised Learning is the process of learning based on known examples which have both an input and a labeled output. Using the data and labels provided in the training dataset, models can learn a prediction function or mapping, which may then be applied to unseen test data. As we discussed in the class, some common methods for supervised learning include the K-Nearest-Neighbors Classifier (KNN) and Naive Bayes.

### Supervised Learning on Text Data

For this assignment, we'll explore kNN and Naive Bayes further and look at different feature selection strategies. First of all, download the dataset for this assignment from Canvas and unpack into an "Assignment 3" directory. 

The dataset is from the RCV1 v2 corpus which is a collection of news articles and category labels. There are 103 categories, and each document can lie in one or more categories. 

The sparse train and test matrices are encoded as nnz x 3 integer matrices, where nnz is the total number of words in all docs. Each row of these matrices contains (doc, word, count) triples. We will store those in sparse matrices whose dimensions are ndocs x nwords

### Application

Your assignment is to perform supervised learning using both the KNN and Naive Bayes classifiers, and evaluate the results. You will also perform multiple types of text feature selection to choose subsets of the total features to use for classification, and evaluate these results as well.

In [26]:
import numpy as np
import scipy.sparse as sp
import matplotlib.pyplot as plt
%matplotlib inline

Now read the data files. 

In [27]:
iwords = np.loadtxt("data/words.imat.txt",dtype=np.int32,)          # training data matrix in nnz x 3 form - rows are (doc, word, count) triples
cats = np.loadtxt("data/cats.imat.txt",dtype=np.int32)             # training labels in an ndocs x ncats matrix
tiwords = np.loadtxt("data/testwords.imat.txt",dtype=np.int32)     # test data matrix in nnz x 3 form
tcats = np.loadtxt("data/testcats.imat.txt",dtype=np.int32)        # testing labels in an ndocs x ncats matrix

iwords and tiwords encode the sparse train and test matrices. Each row of these matrices contains (doc, word, count) triples. We will store those in sparse matrices whose dimensions are ndocs x nwords. The exact matrix type is CSR which stands for compressed sparse rows. Look up its description here: https://en.wikipedia.org/wiki/Sparse_matrix

In [28]:
fwords = iwords[:,2]
train = sp.csr_matrix((fwords, (iwords[:,1], iwords[:,0])))
ndocs = train.shape[0]
nwords = train.shape[1]

test = sp.csr_matrix((tiwords[:,2], (tiwords[:,1], tiwords[:,0])),shape=(4000,nwords))  # need to match the number of cols (words)
ntdocs = test.shape[0]

Although the dataset has 103 category labels, we'll focus on training a single category. We will use Category 6, which is fairly evenly balanced between positives and negatives. 

In [29]:
cat6 = cats[:,6]
tcat6 = tcats[:,6]

# Part 1: KNN Classification

This time we'll train Scikit-Learn's builtin kNN classifier. 

In [30]:
from sklearn.neighbors import KNeighborsClassifier
knnc = KNeighborsClassifier(n_neighbors=3)

And then "fit" it to the data. Notice how fast this is! Of course, training kNN is basically a no-op, and predicting is where all the work lies:

In [31]:
knnc.fit(train, cat6)

KNeighborsClassifier(n_neighbors=3)

Now let's try predicting. This will actually take some time, but hopefully only a few tens of seconds...

In [32]:
preds = knnc.predict(test);preds

array([0, 1, 1, ..., 0, 0, 0], dtype=int32)

Now we can compute the accuracy on the actual test set labels. 

In [33]:
np.mean(tcat6 == preds)

0.82225

Not bad, but kNN is very sensitive to the distance function as we discussed in class. Let's try first normalizing each row of "train" and "test" by the L2-norm of that row, i.e. we will divide by the square root of the sum of the squared word counts in the row. Comparing the normalized documents should do much better. 

Remember, the L2-Norm for a vector $x$ can be written as:

$$  \lVert x \rVert_2  = \sqrt{x_1^2 + x_1^2 + \cdots + x_n^2}$$

## Q1 [20 points]: 

Create two new variables `trainnmat` and `testnmat` which contain the L2-normalized versions of `train` and `test` respectively.

L2-normalization should be performed across each row $r$ of the `train` and `test` matrices, so that:

$$ \sqrt{r_1^2 + r_2^2 + \cdots + r_n^2} = 1 $$

However, you must compute the L2-normalized matrices using `numpy` and `scipy.sparse` operations, using the scikit-learn `normalize` function to generate your answer is not valid.
Also ensure that your output matrix is still sparse. Many NumPy operations will either fail or convert CSR matrices to dense matrices. We prefer to keep this matrix in CSR format for the duration of this lab.

The documentation for the following sparse matrix function may be useful for you:

- `scipy.sparse.csr_matrix` [Docs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html)
- `scipy.sparse.csr_matrix.sum` [Docs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.sum.html)
- `scipy.sparse.csr_matrix.power` [Docs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.power.html#scipy.sparse.csr_matrix.power)
- `scipy.sparse.spdiags` [Docs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.spdiags.html)

Above we have already imported `scipy.sparse` as `sp`.

**HINT:** For both `train` and `test` you need to structure your normalization as a matrix multiply operation. Start by creating a diagonal matrix where the squared sums of each row are along the diagonal elements.


In [36]:
#trainnmat = <your code here>
#testnmat = <your code here>

from sklearn import preprocessing
from numpy.linalg import inv
from scipy.spatial.distance import pdist, squareform
norm_train= train.multiply(train).sum(1)
testnorm = test.multiply(test).sum(1)
trainf = 1.0 / np.sqrt(np.max(norm_train,1))
testf = 1.0 / np.sqrt(np.max(testnorm,1));trainf.shape
trainnmat = sp.spdiags(np.transpose(trainf), 0, ndocs, ndocs) * train
testnmat = sp.spdiags(np.transpose(testf), 0, ntdocs, ntdocs) * test
testnmat.shape

(4000, 31331)

In [37]:
### TEST CASES - DO NOT MODIFY
# These test cases check that your normalization is close (within numerical precision) of the answer
# from sklearn's normalize function.
from sklearn.preprocessing import normalize
skl_norm_train = normalize(train)
skl_norm_train.sort_indices()

skl_norm_test = normalize(test)
skl_norm_test.sort_indices()

trainnmat.sort_indices()
testnmat.sort_indices()

assert(np.allclose(trainnmat.data,skl_norm_train.data))
assert(np.allclose(testnmat.data,skl_norm_test.data))

Now train and evaluate a knn model using the normalized datasets:

In [38]:
knnc.fit(trainnmat, cat6)
preds = knnc.predict(testnmat);preds
np.mean(tcat6 == preds)     

0.90275

## Q2 [10 points]: 
What is the accuracy of your kNN classifier on the normalized data?

In [39]:
#the accuracy of kNN classifier on the normalized data is 0.90275

# Part 2: Naive Bayes Classification

Let's create a Multinomial Naive Bayes classifier. The multinomial distribution models each word in a document as being drawn from a category-specific probability distribution over the vocabulary. 

In [40]:
from sklearn.naive_bayes import MultinomialNB
mnb = MultinomialNB()

Now we can fit the NB model, which simply means tabulating the word counts for each label.

In [41]:
mnb.fit(train, cat6)

MultinomialNB()

...and predict the category labels for cat6:

In [42]:
preds = mnb.predict(test);preds

array([0, 1, 1, ..., 1, 0, 1], dtype=int32)

## Q3 [10 points]: 

Let's check the accuracy below. How did NB do compared to kNN? 

In [43]:
np.mean(tcat6 == preds)

0.8975

In [44]:
#TODO: Answer this question here
#The accuracy is better with kNN comapres to the NB beacuse kNN had the accuracy of 0.90275

# Part 3: Frequency-Based Feature Selection

We argued that frequency-based feature selection is useful approach to culling features to reduce the size of a model. In some cases, accuracy improves as well. 

Let's construct a matrix rcounts which contains the counts of features from the training set. 

In [45]:
rcounts = (train>0).sum(0)
rcounts

matrix([[3728, 4438, 3488, ...,    1,    1,    1]])

In [46]:
rcounts.size

31331

Next we'll find the indices of non-zero elements of the predicate (rcounts >= count). These are the features that occurred at least count times. ii is the indices of these features.

We then extract the columns of the train and test matrices corresponding to those high-count features. The last line tells us how many features we kept.

In [47]:
threshold = 2
ii0 = np.asarray(np.nonzero(rcounts >= threshold)[1])
ii = ii0.reshape(ii0.size,)

strain = train[:,ii]
stest = test[:,ii]
ii.size

18330

Finally we run training and evaluation on train/test data restricted to those columns to get an accuracy score.

In [48]:
mnb.fit(strain, cat6)
preds = mnb.predict(stest);preds
np.mean(tcat6 == preds)

0.8955

## Q4 [20 points]
For the following question, use the `threshold` variable set above.

For threshold counts of 1, 2, 5, 10, 20, tabulate the count, number of features kept, and the accuracy score. 

In [49]:
#TODO: Answer the question here
import pandas as pd

mnb.fit(strain, cat6)
preds = mnb.predict(stest);preds
np.mean(tcat6 == preds)

thresholds = [2,5,10,20]
li = [[1, ii.size, np.mean(tcat6 == preds)]]
for t in thresholds:
    threshold = t
    ii0 = np.asarray(np.nonzero(rcounts >= threshold)[1])
    ii = ii0.reshape(ii0.size,)
    strain = train[:,ii]
    stest = test[:,ii]
    ii.size
    mnb.fit(strain, cat6)
    preds = mnb.predict(stest);preds
    np.mean(tcat6 == preds)
    li.append([t, ii.size, np.mean(tcat6 == preds)])


df3 = pd.DataFrame(li,columns=['Count','Features', 'Score'])
df3

Count  Features    Score
0      1     18330  0.89550
1      2     18330  0.89550
2      5      9632  0.89000
3     10      6169  0.88675
4     20      4046  0.88400

## Q5 [10 points]

How did frequency-based feature selection affect accuracy? How much was the model size (proportional to words kept) reduced? 

(`ii.size` is the reduced model size, `rcounts.size` is the original model size.)

In [50]:
#TODO: Answer the question here
#The more features yielded higher accuracy than lower number of features. Model size was reduced by almost half each time.

# Part 4: Feature Selection Based on Statistical Tests

Let's try a more sophisticated method to select the (word) features. Since we have count data, the Chi-Squared test is appropriate. Remember that Chi-squared can be used on 2x2 contingency tables to determine if there is an "interaction". In this case, we test if the label interacts with the feature. 

In [51]:
fcounts = np.zeros((4,nwords))           # holds the counts of various combinations of label and feature
expected = np.zeros((4,nwords))          # holds the predicted counts of those same combinations

In [52]:
mcat6 = cat6.reshape(1,ndocs)
fcounts[0,:] = mcat6 * (train > 0)                                   # label and word both = 1
fcounts[1,:] = (1-mcat6) * (train > 0)                               # label = 0, word = 1
fcounts[2,:] = np.sum(mcat6) * np.ones((1,nwords)) - fcounts[0,:]    # label = 1, word = 0
fcounts[3,:] = np.sum(1-mcat6) * np.ones((1,nwords)) - fcounts[1,:]  # label = 0, word = 0

Review the definition of the CHI-squared test here:
https://en.wikipedia.org/wiki/Chi-squared_test .

In [54]:
allf = np.sum(fcounts,axis=0)
pword0 = (fcounts[2,:] + fcounts[3,:]) / allf    # prob word = 0
pword1 = (fcounts[0,:] + fcounts[1,:]) / allf    # prob word = 1
pcat0 = (fcounts[1,:] + fcounts[3,:]) / allf     # prob label = 0
pcat1 = (fcounts[0,:] + fcounts[2,:]) / allf     # prob label = 1

If label and word occurence are independent, then the expected count should be the product of the total count and the two corresponding probabilities:



## Q6 [20 points]: 

Implement the expected counts here. 

**HINT** This can be computed entirely with the variables from the previous code cell, with no other variables needed.

In [55]:
#TODO: Implement this
expected[0,:] = allf * pcat1 * pword1  
expected[1,:] = allf * pcat0 * pword1 
expected[2,:] = allf * pcat1 * pword0  
expected[3,:] = allf * pcat0 * pword0 

To compute the chi-squared statistic value, we want the sum of the squared differences between actual and predicted counts, normalized by the expected counts. The following two cells perform that calculation. We have to guard against divide-by-zero so we use a small lower-bound on the denominator.

In [56]:
diff = fcounts - expected
safeexpected = np.maximum(expected,1e-7)

In [57]:
chisquaredstat = np.sum(np.multiply(diff, diff) / safeexpected, axis=0)
chisquaredstat.shape

(31331,)

In [58]:
chisquaredstat

array([ 83.64630905,  97.8608861 , 660.13078932, ...,   0.8417048 ,
         0.8417048 ,   0.8417048 ])

We can then use the chi-squared statistic to choose the best features. The CHI-squared stastistic is a measure of interaction between a word feature and the label. The higher its value, the better that word should be as a predictor of the category. We set a threshold and take the words whose chi-squared statistic is above this value:

In [59]:
chisqthreshold = 0.001
ii0 = np.asarray(np.nonzero(chisquaredstat > chisqthreshold)[0])
ii = ii0.reshape(ii0.size,);ii.shape

(31281,)

Finally we use the indices of good features selected by the chi-squared filter to train and evaluate the Naive Bayes model.

In [60]:
strain = train[:,ii]
stest = test[:,ii]
mnb.fit(strain, cat6)
preds = mnb.predict(stest);preds
np.mean(tcat6 == preds)

0.898

## Q7 [10 points]

Using a chi-sqaured critical value table (one can be found here: [Link](https://people.richland.edu/james/lecture/m170/tbl-chi.html)), find the table row for 1 degree of freedom (`df=1`), and set `chisqthreshold` to a few different values from the table. Record your accuracies here.

In a few sentences describe how well this feature selection process works on the test data.


Answer :
chisqthreshold -> accuracy 
0.005 -> 0.89775
0.01 -> 0.8975
0.025 -> 0.89725
0.10 -> 0.8937
0.90 -> 0.8945
0.95->0.895
0.975 -> 0.8947
0.995 -> 0.89475

The feature selection works well with accuracy near 0.8 -0.9. But I noticed that when the threshold increases, the accuracy decreases. This maybe beacuse some of the chi-square stats are above that certain threshold. 

## Submit

Submit your Assignment 3 notebook on Canvas.